Replace quad_bilinear_interp with inverse distance weighting in quad_utils.

hkershaw-brown opened the pull request:
## Description:
<!--- Describe your changes -->
The current quad_utils_mod interpolation method (bilinear) can give out-of-bounds values for some quads. 
This problem was originally reported by Nuo Chen in CICE #833 when using the QCEFF.  
The CICE bilinear interpolation was replaced by inverse distance weighting in #848.  This is the same code as quad utils. Note, replacing the CICE interpolation with quad utils is not done in this pull request. 

There is a notebook to call dart from python in https://github.com/hkershaw-brown/DART-probe/blob/main/ice-interp.ipynb if you want to reproduce the interpolation plots (see [README](https://github.com/hkershaw-brown/DART-probe/blob/main/README.md) for dart library build).  

### Fixes issue
<!--- link to github issue(s) -->
Fixes #860
Fixes #453 since no rotation needed in the IDW. See #833 for plots of rotations

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?
<!-- Put an `x` in all the boxes that apply: -->
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

There is no documentation for quad utils at the moment. 

### Tests
This is a draft at the moment.
Questions to answer: 
bitwise differences for which models - who cares, it is a bug fix vs. loads of people care. 

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

hkershaw-brown committed:
fix: replace bilinear interpolation with inverse distance weighting in quad_utils_mod

fixes #860
Same fix as #833 #848 that was for cice model_mod only.
Removed the rotation code - see #453 - does not fix the out-of-bounds
 and is also rotating in the oppostite direction. Not used in IDW.

removed svn detritus.

Using VERTISUNDEF to get horizontal only distances. I am not sure if
this will need to be modififed for other location_mods

hkershaw-brown committed:
compile quad_utils_mod only model with threed_sphere

hkershaw-brown committed:
compile observation convters with quad_utils only for threed_sphere

The converters in the repo all build succesfully without this change,
because they all use threed_sphere location_mod. However maybe there
are some converters in the wild that do not use threed_sphere.

hkershaw-brown commented:
build everything action: https://github.com/NCAR/DART/actions/runs/15426138267 caveat GEOS (Helen are you ever going to fix this? no).

jlaucar reviewed (commented):
This looks correct to me. It will definitely not bitwise reproduce. It remains unclear how often the result could be out of the range of the corner values due to roundoff. It we run into applications where this occurs, we may have to remove the error failure so that the currently unused bounds enforcement code comes into play.

jlaucar reviewed (changes requested):
This code still looks good to me. However, the comment stating that it is 'unclear' whether round-off could lead to the interpolated value being outside of the range of the vertex values is now clear. The modified code I'm using in aether_cube_sphere occasionally has an interpolated value outside or range when all of the vertex values are identical to machine precision. This comment should probably be changed to
"Round-off can lead to result being outside of range of grid points"

The comment further down should also be changed since we are issuing a MSG rather than an ERR. The part that says "this will not happen with current error check" should be deleted.

jlaucar committed:
If all four point values are identical in interpolate, just return
that value. Also cleaned up the comments on what happens if the
interpolated value is outside the range of the point values.

hkershaw-brown committed:
Replace exact equality fp check with tolerance-based comparison in quad_idw_interp

hkershaw-brown committed:
remove svn relics

Make the error message about update_reg_list slightly more helpful
https://github.com/NCAR/DART/pull/887#issuecomment-3382582285

hkershaw-brown committed:
fix: round off message on all tasks, since it may occur on tasks other than 0

note E_ALLMSG does not get written to the log file, only std out.

hkershaw-brown committed:
comment: remove bilinear interp comments, now idw

hkershaw-brown commented:
also check pole for IDW in addition to #978

hkershaw-brown commented:
Note Moha working on improved quad algorithms.

hkershaw-brown commented:
closing this pull request as it is now, pull #1089 has a more stable method than Cramer's rule method. 
If idw goes in to quad_utils_interp, go in as an optional interpolation method.
